sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/util/CharVarcharCodegenUtils.java

```diff
@@ -47,19 +47,23 @@ public static UTF8String charTypeReadSideCheck(UTF8String inputStr, int limit) {
   }
 
   public static UTF8String varcharTypeWriteSideCheck(UTF8String inputStr, int limit) {
-    if (inputStr != null && inputStr.numChars() <= limit) {
-      return inputStr;
-    } else if (inputStr != null) {
-      // Trailing spaces do not count in the length check. We need to retain the trailing spaces
-      // (truncate to length N), as there is no read-time padding for varchar type.
-      // TODO: create a special TrimRight function that can trim to a certain length.
-      UTF8String trimmed = inputStr.trimRight();
-      if (trimmed.numChars() > limit) {
-        throw new RuntimeException("Exceeds varchar type length limitation: " + limit);
-      }
-      return inputStr.substring(0, limit);
-    } else {
+    if (inputStr == null) {
       return null;
+    } else {
+      int numChars = inputStr.numChars();
+      if (numChars <= limit) {
+        return inputStr;
+      } else {
+        // Trailing spaces do not count in the length check. We need to retain the trailing spaces
+        // (truncate to length N), as there is no read-time padding for varchar type.
+        int tailSpacesMayNeedTrim = numChars - limit;
+        int lengthAfterTrim = inputStr.effectiveLength(tailSpacesMayNeedTrim);
+        if (lengthAfterTrim == limit) {
```

Comment:
This isn't right. `lengthAfterTrim` is the length of bytes, while `limit` is the length of chars.

Reply:
oh...